File: python/BOJ1697.py
from collections import deque
import sys

input = sys.stdin.readline
# https://www.acmicpc.net/problem/1697

N, K = [int(a) for a in input().rstrip().split()]

"""
모든 위치별로 최단 도달 시간을 저장한다.
방법은 bfs 로 간선은 갈 수 있는 점이 된다.
만약 저장된 최소 시간이 더 짧으면 생략
"""

dist = [999999] * 100001

q = deque()
visit = [False] * 1000001


def bfs():
    q.append((N, 0))
    visit[N] = True
    dist[N] = 0

    while q:
        x, d = q.popleft()

        for ax in [x - 1, x + 1, x * 2]:
            if ax < 0 or ax > 100000 or visit[ax] or dist[ax] <= d + 1:
                continue
            q.append((ax, d + 1))
            visit[ax] = True
            dist[ax] = d + 1


bfs()
print(dist[K])

File: java/BOJ/BOJ1697.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class BOJ1697 {
    // https://www.acmicpc.net/problem/

    // 변수 설정
    static FastReader fr = new FastReader();
    static int N, K;
    static boolean[] visit;
    static int[] dist;

    // 입력 함수
    static void input(){
        N = fr.nextInt();
        K = fr.nextInt();
        visit = new boolean[100001];
        dist = new int[100001];
    }

    static void bfs(){
        Queue<Integer> q = new LinkedList<>();
        q.add(N);
        visit[N] = true;
        dist[N] = 0;

        int[] dir;

        while(!q.isEmpty()){
            int x = q.poll();
            if(x == K) {
                System.out.println(dist[x]);
                break;
            }

            dir = new int[]{x - 1, x + 1, x * 2};
            for(int next : dir){
                if(next >= 0 && next <= 100000 && !visit[next]){
                    q.add(next);
                    visit[next] = true;
                    dist[next] = dist[x] + 1;
                }
            }
        }
    }

    public static void main(String[] args) throws Exception {
        input();
        bfs();
    }

    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        public FastReader(String s) throws FileNotFoundException {
            br = new BufferedReader(new FileReader(new File(s)));
        }

        String next() {
            while (st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }

        long nextLong() {
            return Long.parseLong(next());
        }

        double nextDouble() {
            return Double.parseDouble(next());
        }

        String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            } catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }
}